### ISSUE TITLE
F821 (Undefined name) and F722 (bad forward annotation) can trigger on `@no_type_check`-ed annotations

### ISSUE THREAD

**Issue by Wuestengecko** [astral-sh/ruff#13824]
Hi! I found that F821 ("Undefined name") and F722 ("Syntax error in forward annotation") can erroneously trigger on functions that marked `@no_type_check`, if they happen to either use names that aren't in scope, or aren't even valid Python.

Consider the following example file:

```python
from typing import no_type_check

@no_type_check
def f821(arg: "A") -> "R":
    pass

@no_type_check
def f722(arg: "this isn't python") -> "this isn't python either":
    pass
```

Save it as "demo.py" and run ruff with:

    ruff check --isolated --select=F722,F821 demo.py --output-format concise

This will emit the following errors, which are all false-positives:

```
demo.py:4:16: F821 Undefined name `A`
demo.py:4:24: F821 Undefined name `R`
demo.py:8:15: F722 Syntax error in forward annotation: `this isn't python`
demo.py:8:39: F722 Syntax error in forward annotation: `this isn't python either`
Found 4 errors.
```

I'm presuming (but haven't tested) that this could apply to other rules that inspect annotations as well. Ruff should treat these annotations as opaque strings without any meaning. As per [PEP-484](https://peps.python.org/pep-0484/#the-meaning-of-annotations):

>  Functions with the `@no_type_check` decorator should be treated as having no annotations.

---

**ruff --version**: 0.7.0
**settings**: no config file exists

**Comment by MichaReiser** [comment]
Thanks for reporting. Suppressing some rules annotated with `@no_type_check` makes sense to me.

**Comment by AlexWaygood** [comment]
This seems reasonable to me too

**Comment by charliermarsh** [comment]
Should we just skip visiting them altogether when in a `@no_type_check` context?

**Comment by AlexWaygood** [comment]
> Should we just skip visiting them altogether when in a `@no_type_check` context?

That might make sense, yeah

**Comment by MichaReiser** [comment]
> Should we just skip visiting them altogether when in a @no_type_check context?

Are you suggesting skipping such functions only for `F821` and `F722` or all rules?

**Comment by charliermarsh** [comment]
I'm suggesting skipping looking at those nodes entirely, for all rules.

**Comment by charliermarsh** [comment]
Alternatively, we could visit them as strings (so enforce rules like implicit concat -- weird but possible) but skip parsing them and visiting the _content_.

**Comment by MichaReiser** [comment]
To clarify, with all nodes you mean skipping the entire function or only the type annotations? I'm somewhat fine skipping the annotations but would find it a stretch if we skip the entire function

**Comment by charliermarsh** [comment]
Just the annotations. The rest of the function would be unchanged.

**Comment by MichaReiser** [comment]
We'll revert the change. It's a bit more involved, see https://github.com/astral-sh/ruff/issues/14698
